### ISSUE TITLE
`pub type` type aliases have unclear public API participation, with downstream impact on trait-sealing

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#1254]
> The issue reported by @beeb is much more challenging, and is due to our very limited ability to analyze `pub type` items. We have several more bugs related to `pub type`, but this is by far the most severe and it will be a difficult one to fix. I wrote out a sketch of some ideas [in a draft PR](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/825), but there's a long way to go before that PR can be merged.
> 
> To get some intuition about why `pub type` is full of edge cases, consider a case like `pub type Foo = Bar<Quux>` and try to figure out the rules about when `Foo, Bar, Quux` are public API or not (and whether they would cause trait-sealing), depending on whether any one or more of them are `#[doc(hidden)]`, private, or public but inside a private module and therefore non-importable. You'll quickly see that the relationship is very complex! For example, `Bar` could be non-importable by itself (so use of `Bar` in a trait method would cause the trait to be sealed), _except_ if the `Bar` in the trait in question is actually `Bar<Quux>` specifically — in which case that unseals the trait by allowing use of the type alias `Foo` instead.

 _Originally posted by @obi1kenobi in [#1200](https://github.com/obi1kenobi/cargo-semver-checks/issues/1200#issuecomment-2817633397)_

See [the comments in this draft PR](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/825/files#diff-b077885a56f85cf6c81b79fa7ada6d25ec11dd83e4bed6d251b8e832f76eb69c) for more context on the difficulties caused by type aliases, including cases of non-public API becoming public API due to type aliases, and traits reversing sealing status in ways that require fine-grained tracking not just of type aliases but the type contained within.

Solving this problem properly will take hundreds, possibly thousands of hours of work. We will eventually get there, but I don't have a timeline to offer. I'd encourage everyone interested in speeding up our tempo there to [become a GitHub Sponsor of the project](https://github.com/sponsors/obi1kenobi)!

**Comment by beeb** [comment]
Thanks for creating the issue and investigating!

**Comment by kornelski** [comment]
There may be some unclear intention from crate authors, e.g. maybe a crate author had `pub type Token = ()` and wanted users to use `Token`, and considers a direct use of `()` cheating. But that's not how Rust works — type aliases are not newtypes.

Strictly from language's perspective, `type` aliases are transparent, so when you need to know whether a trait is sealed because something like `method(_: Token)` contains an unnameable type, you need ability to find *types* that can be named and are compatible with the `Token`'s underlying *type*.

I think the surprising difficulty appeared here due to [lack of type support](https://github.com/obi1kenobi/cargo-semver-checks/issues/149) in `cargo-semver-checks`. You can't reason about type aliases directly as their own things. This just obscures the problem, like trying to do multiplication using roman numerals. You need the necessary building blocks of name resolution and type normalization. Unfortunately, the general problem of checking whether an argument seals a trait is really a problem of type equivalence, and that can get [as complex as types can](https://rust-lang.github.io/chalk/book/). `method(_: <T::Output<'static> as Trait<U>>::Assoc<Self>)` makes type aliases look really easy! ;)

**Comment by kornelski** [comment]
I think an 80/20 solution is possible. I've outlined an algorithm for it below. It may seem long, but it's just a couple of not-so-complicated recursive algorithms. It's similar to what `rustdoc` does internally to display types behind private aliases and to create canonical URLs for types.

Terminology: by item being *accessible* I mean there exists some *path* that can be used from another crate to name this item. When any component of a path is not accessible, that path is not accessible (but it doesn't prevent the item it refers to from being accessible via some alternative path). When evaluating whether a path still exist in the public API, treat any `#[doc(hidden)]` items as not accessible. When evaluating whether old paths still work in the current version, don't let `#[doc(hidden)]` affect accessibility. 

## Resolution/Normalization

Convert accessible type aliases to the most basic types possible, and save the fact that the resolved type is accessible.

When you have:

```rust
pub type Alias = Bar<Quux>
```

1. If the `Alias` itself is not accessible *at this path* (e.g. it's in a private module), then there's nothing more to do. Stop here. If there's another path that is accessible and happens to use the `Alias`, then that path will take care of it.

2. Completely resolve the right side, following all further type aliases, `use`s etc. Ignore visibility/accessibility when doing so! Follow private modules, private `use`s, private type aliases, associated types, extern crate, use `#[doc(hidden)]` items, everything. This code has compiled, so it all has the right to be there. I think `rustdoc` should be helping you here by already replacing undocumented/unnameable paths with equivalent public ones where possible, but if you run into dead-ends, you'll need more info from rustdoc.

If it was:

```rust
mod private { #[doc(hidden)] pub use u32 as int; }
struct X;
type Quux = (private::int, X);
use Vec as Bar;
pub type Alias = Bar<Quux>
```

after resolution you'd get:

```rust
pub type Alias = Vec<(u32, X)>;
```

3. Make a note that the fully resolved type `Vec<(u32, X)>` is accessible (since having access to the `Alias` is identical to having access to this type and all other aliases of this type).

## Type equivalence

When you have to check if something that isn't obviously accessible can be named anyway, resolve the name, then check if the resolved type as a whole can be named, or whether all parts of the type can be named, recursively:

```rust
fn maybe_sealed(_: Token)
```

Resolve the `Token` type like in step 2, all the way down to its leafmost constituent types. For example, the `Token` could end up being something like `[Vec<(u32, X)>; 2]` through completely different type aliases:

```rust
pub type Token = Two<Vec<Pair>>;
type Two<T> = [T; 2];
type Pair<T = X> = (u32, T);
```

Then check if the resolved `[Vec<(u32, X)>; 2]` has been seen as accessible as a whole alias. If not, then check all of its parts. `[T; 2]` is accessible by being a built-in type, and then `T == Vec<(u32, X)>` is accessible, because we've seen that type in the step 3 before. If the previous `Alias` wasn't accessible, then you'd keep checking for `Vec`, `(u32, X)`, and then `u32` and `X`.

## Generics and associated types

Fortunately, `type` aliases don't support generic bounds — `type A<T: Bound> = T` is not legal. Unfortunately, there can still be trait resolution involved, e.g. `type X<U> = <U as Add>::Output`. Traits and their associated types can get infinitely complex, so I don't recommend touching that. Type comparison can still be useful if it sometimes answers "maybe"!

Default arguments can be brute-forced by considering them separately with and without the default, e.g.

```rust
pub type Alias<T = Foo, U = Bar> = (T, U);
```

conceptually can be simplified to resolving separate aliases for `(Foo, Bar)`, `(Foo, U)`, `(T, Bar)`, `(T, U)` (avoiding 2^n cost is an implementation issue :)

```rust
use Vec as Bar;
pub type Alias<T> = Bar<T>;
```

This resolves to `Vec<T>`, and unfortunately it's necessary to keep the generic argument. Later when you want to check if this makes `Vec<(u32, X)>` accessible, you'll need a bit of fiddly logic to check that `Vec`s match, so `T == (u32, X)`, and being equal to `T` means that `(u32, X)` must be accessible on its own.

```
type Alias<T> = Vec<(T, T)>;
```

This starts requiring showing that `u32` != `X` for `Vec<(u32, X)>`, and that needs to become an equation solver, rather than a simple recursive comparison. At this point I'd make it return "maybe".
